src/AvoDeduplicator.ts

```diff
@@ -0,0 +1,183 @@
+import { AvoSchemaParser } from "./AvoSchemaParser";
+import { deepEquals } from "./utils";
+
+export class AvoDeduplicator {
+  avoFunctionsEvents: { [time: number]: string } = {};
+  manualEvents: { [time: number]: string } = {};
+
+  avoFunctionsEventsParams: {
+    [eventName: string]: { [propName: string]: any };
+  } = {};
+  manualEventsParams: { [eventName: string]: { [propName: string]: any } } = {};
+
+  shouldRegisterEvent(
+    eventName: string,
+    params: { [propName: string]: any },
+    fromAvoFunction: boolean
+  ): boolean {
+    this.clearOldEvents();
+
+    if (fromAvoFunction) {
+      this.avoFunctionsEvents[Date.now()] = eventName;
+      this.avoFunctionsEventsParams[eventName] = params;
+    } else {
+      this.manualEvents[Date.now()] = eventName;
+      this.manualEventsParams[eventName] = params;
+    }
+
+    let checkInAvoFunctions = !fromAvoFunction;
+
+    return !this.hasSameEventAs(eventName, params, checkInAvoFunctions);
+  }
+
+  private hasSameEventAs(
+    eventName: string,
+    params: { [propName: string]: any },
+    checkInAvoFunctions: boolean
+  ): boolean {
+    let result = false;
+
+    if (checkInAvoFunctions) {
+      for (const otherEventName in this.avoFunctionsEventsParams) {
+        if (this.avoFunctionsEventsParams.hasOwnProperty(otherEventName)) {
+          if (otherEventName === eventName) {
+            const otherParams = this.avoFunctionsEventsParams[eventName];
```

Comment:
Is there any change this would return undefined?

Reply:
there always is :)

Reply:
how would we handle it?

Reply:
somehow handle the case where otherParams is undefined below, probably something like

```
otherParams && deepEquals(params, otherParams)
```